### PR TITLE
Compute pipelines, transition dependencies, fixes

### PR DESCRIPTION
**Pull request by jkunstwald**
## General
- Add compute pipeline, cmd::dispatch
- Add root constants to compute and graphics
- Add comprehensive blend state support to graphics pipelines
- Add cmd::transition_image_slices for subresource transitions
- Add cmd::copy_texture
- Reduce cmd::copy_buffer_to_texture requirements
- Several small API changes relating to compute shaders
- Add Backend::free_range, Backend::flushMappedMemory
- Add shader_domain_bits, explicit dependency field to transition commands
- Add scissor rectangle and vertex + index offset to cmd::draw
- Remove initial state argument from buffer creation, add "allow UAV" instead
- Add cmd::debug_marker, mapping to PIX labels and VK_EXT_debug_utils labels
- Add backend API for programmatic render diagnostic capture triggering
    (PIX, NSight, RenderDoc)
- Add Renderdoc app header
- Add convenience methods to most cmd:: types
- Add native feature flags to backend config (currently: Vulkan API dump)
- Remove unrepresentable pr formats, add format::rg8un, format::r8un
- Fix resize crashes
- cc::hash

## D3D12
- Fix UAV cubemaps
- Fix SRV structured buffers
- Expand DRED output
- Add PIX runtime dependency, PIX and Renderdoc capturing
- Fix redundant rebinds in cmd::draw
- Fix crashes with enabled GBV on Intel GPUs
- Fix crashes with enabled DRED when running inside NSight
- Fix crashes when using PSOs with zero shader arguments
- Fix crashes when (non-range) freeing an invalid resource handle
- IDXGI queue message filter for GBV startup message
- Use non-wchar debug names with 1903 `WKPDID_D3DDebugObjectName` introduction

## Vulkan
- Fix spurious driver crash in compute pipeline (previously mistaken for a driver bug, https://developer.nvidia.com/nvidia_bug/2786273)
- Add GPU-assisted validation, reserved diagnostic descriptor set, and best practices layer (for validation_level::on_extended)
- Fix redundant pipeline stage dependencies
- Fix incomplete pipeline stage dependencies and draw/dispatch based cache
- Fix multi-descriptor bindings
- Fix cubemap UAVs only writing to X+
- Fix SPIR-V descriptor range merging
- Enable anisotropy and geometry shaders
- Add Renderdoc capturing, fix crashes when loading into renderdoc with enabled validation
- Add reflection vs. argument shape consistency check
- Fix crash when unsuitable GPUs are present

### CODE REVIEW

File: src/phantasm-renderer/backend/Backend.hh [hunk 1]
```diff
@@ -39,12 +39,15 @@ public:
     /// causes an internal resize
     virtual void onResize(tg::isize2 size) = 0;
 
-    /// gets the current backbuffer size
+    /// returns the current backbuffer size
     [[nodiscard]] virtual tg::isize2 getBackbufferSize() const = 0;
```
Comment (Philip-Trettner) [inline src/phantasm-renderer/backend/Backend.hh]: hm I feel like adding [[nodiscard]] for EVERY non-void method is overkill and introduces way more noise into the code than necessary. maybe we should only add it when the return value is important but the method name does not make it clear? I mean for `getXYZ` it is obvious that you need the return value but for `clearPendingSize()` it totally is not.

File: src/phantasm-renderer/backend/assets/vertex_attrib_info.hh [hunk 2]
```diff
@@ -51,35 +51,27 @@ constexpr format to_attribute_format()
...
-    else if constexpr (tg::is_comp_like<T, 3, cc::uint16>)
-        return af::rgb16u;
     else if constexpr (tg::is_comp_like<T, 2, cc::uint16>)
         return af::rg16u;
     else if constexpr (tg::is_comp_like<T, 1, cc::uint16>)
         return af::r16u;
 
     else if constexpr (tg::is_comp_like<T, 4, cc::int8>)
         return af::rgba8i;
-    else if constexpr (tg::is_comp_like<T, 3, cc::int8>)
```
Comment (Philip-Trettner) [inline src/phantasm-renderer/backend/assets/vertex_attrib_info.hh:72]: why are 3-comp formats removed?

File: src/phantasm-renderer/backend/command_stream.hh [hunk 3]
```diff
@@ -93,46 +120,235 @@ PR_DEFINE_CMD(transition_resources)
...
+    handle::pipeline_state pipeline_state = handle::null_pipeline_state;
+    handle::resource vertex_buffer = handle::null_resource;
+    handle::resource index_buffer = handle::null_resource;
+    unsigned num_indices = 0;
+    unsigned index_offset = 0;
+    unsigned vertex_offset = 0;
+
+    /// the scissor rectangle to set, none if -1
+    /// left, top, right, bottom of the rectangle in absolute pixel values
+    tg::ivec4 scissor = tg::ivec4(-1, -1, -1, -1);
```
Comment (Philip-Trettner) [inline src/phantasm-renderer/backend/command_stream.hh]: `iaabb2`?

File: src/phantasm-renderer/backend/assets/vertex_attrib_info.hh (hunk 2 above)
Comment (jkunstwald) [reply src/phantasm-renderer/backend/assets/vertex_attrib_info.hh:72]: 3-component formats do not exist in D3D12 and are only rarely supported in Vulkan (for example, rgb16f 8.5% vs rgba16f 100%, https://vulkan.gpuinfo.org/listformats.php)

File: src/phantasm-renderer/backend/Backend.hh (hunk 1 above)
Comment (jkunstwald) [reply src/phantasm-renderer/backend/Backend.hh]: I agree, i removed them for the obvious getters.

File: src/phantasm-renderer/backend/assets/vertex_attrib_info.hh (hunk 2 above)
Comment (Philip-Trettner) [reply src/phantasm-renderer/backend/assets/vertex_attrib_info.hh:72]: okay then let's leave them out of the backend. I'll have to see if we can do something about them in the frontend. For example, Jascha's stuff can build any component structure it wants internally.

**Review by Philip-Trettner — APPROVED** [review]

**Pull request merged**